crates/bounded_planet_core/src/networking/systems.rs

```diff
@@ -117,190 +123,186 @@ pub fn receive_net_events(
-            )).expect("Failed to send network event");
-            return;
-        }
-    };
-
-    // Create a new MPSC which the ECS can use to send packets through this connection
-    let (send, recv) = unbounded_channel();
+}
 
-    // Send an intial event indicating that this connection opened
-    event_sender
-        .send(ReceiveEvent::Connected(cid, send))
-        .expect("Failed to send network event");
+impl Connecting {
+    pub fn new(connecting: quinn::Connecting, event_sender: UnboundedSender<ReceiveEvent>) -> Self {
+        Connecting {
+            id: ConnectionId::new(),
+            connecting,
+            event_sender
+        }
+    }
 
-    // Spawn a task which polls for new incoming streams
-    tokio::spawn(handle_incoming_streams(uni_streams, event_sender.clone(), cid));
+    /// Wait for the connection to finish opening and then spawn the tasks to handle it
+    pub async fn run(self) {
+        info!("connection incoming: {:?}", self.id);
 
-    // Spawn a task which opens new outgoing streams and sends packets to them
-    tokio::spawn(send_to_streams(cid, connection, recv, event_sender));
-}
-
-async fn handle_incoming_streams(
-    mut uni_streams: IncomingUniStreams,
-    event_sender: UnboundedSender<ReceiveEvent>,
-    id: ConnectionId
-) {
-    // Keep getting events from the connection until it closes
-    while let Some(stream) = uni_streams.next().await {
-        match stream {
-            Err(quinn::ConnectionError::ApplicationClosed { .. }) => break,
+        // Wait for connection to finish connecting
+        let quinn::NewConnection { connection, uni_streams, .. } = match self.connecting.await {
+            Ok(connection) => connection,
             Err(e) => {
-                event_sender.send(ReceiveEvent::NetworkError(
+                self.event_sender.send(ReceiveEvent::NetworkError(
                     NetworkError::ConnectionError(e)
                 )).expect("Failed to send network event");
-                break;
-            },
-            Ok(recv) => {
-                tokio::spawn(read_from_stream(id, recv, event_sender.clone()));
+                return;
             }
         };
-    }
 
-    // Send a final event indicating that this connection closed
-    event_sender
-        .send(ReceiveEvent::Disconnected(id))
-        .expect("Failed to send network event");
+        // Create a new MPSC which the ECS can use to send packets through this connection
+        let (send, recv) = unbounded_channel();
+
+        // Send an intial event indicating that this connection opened
+        self.event_sender
+            .send(ReceiveEvent::Connected(self.id, send))
+            .expect("Failed to send network event");
+
+        // Start running tasks to send/receive to this connection
+        tokio::spawn(Connected::new(
+            self.id,
+            connection,
+            uni_streams,
+            recv,
+            self.event_sender,
+        ).run());
```

Comment:
Suggestion in reference to the somwhat unidiomatic usage of this new function. Refer to the struct defintion for more information.

```suggestion
        tokio::spawn(Connected {
            id: self.id,
            connection,
            uni_streams,
            recv,
            send: self.event_sender,
        }.run());
```
Or alternatively:
```suggestion
        let connected = Connected {
            id: self.id,
            connection,
            uni_streams,
            recv,
            send: self.event_sender,
        };
        
        tokio::spawn(connected.run());
```